flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1270,37 +1414,116 @@ def test_simple_transitions(self, two_dataset_states):
         to any ``DesiredDataset``, followed by any other state of the same
         dataset.
         """
+        try:
+            test_objects = BlockDeviceCalculatorTestObjects(self)
+
+            initial_dataset, next_dataset = two_dataset_states
+
+            dataset_id = initial_dataset.dataset_id
+
+            # Set the mountpoint to a real mountpoint in desired dataset states
+            # that have a mount point attribute.
+            mount_point = test_objects.deployer._mountpath_for_dataset_id(
+                unicode(dataset_id))
+            if getattr(initial_dataset,
+                       'mount_point',
+                       _NoSuchThing) is not _NoSuchThing:
+                initial_dataset = initial_dataset.set(mount_point=mount_point)
+            if getattr(next_dataset,
+                       'mount_point',
+                       _NoSuchThing) is not _NoSuchThing:
+                next_dataset = next_dataset.set(mount_point=mount_point)
+
+            # Converge to the initial state.
+            try:
+                test_objects.run_to_convergence([initial_dataset])
+            except DidNotConverge as e:
+                self.fail(
+                    "Did not converge to initial state after %d iterations." %
+                    e.iteration_count)
+
+            # Converge from the initial state to the next state.
+            try:
+                test_objects.run_to_convergence([next_dataset])
+            except DidNotConverge as e:
+                self.fail("Did not converge to next state after %d "
+                          "iterations." % e.iteration_count)
+        finally:
+            test_objects.cleanup_example()
+
+    @given(
+        two_dataset_states=TWO_DESIRED_DATASET_STRATEGY
+    )
+    def test_docker_protection(self, two_dataset_states):
+        """
+        Given an initial empty state, ``BlockDeviceCalculator`` will converge
+        to any ``DesiredDataset``, followed by any other state of the same
+        dataset. During that time, it will never be the case that a
+        successfully write to a mount path that has been sent to the control
+        agent will not be persisted to the backing volume.
```

Comment:
Not quite what we're asserting. Rather "will only be persisted to the backing volume; it might however fail to write if the dataset is not mounted."

Possibly a more clever test would actually distinguish between "Failed to write" and "successfully wrote" based on current dataset state? Cause if it's *always* fail-to-write this test will pass as is, but that's obviously not ideal.